### PR TITLE
feat: TUI logs view

### PR DESCRIPTION
**Pull request by JosephGoulden**
Adding a logs tab to TUI as described in issue #821 

I used different colours for Info/Warn/Error log levels. I didn't make the panics pop up in any way, as described in the issue, but they are logged with full stack trace in red.

I've manually tested stopping/starting grin with and without the TUI, various log levels and logging to a file on and off. All is working well. There are some small code changes around the logging when stop starting the server that came from this testing and I believe all the relevant information is logged in the right places.

image

### CODE REVIEW

File: src/bin/grin.rs [hunk 1]
```diff
@@ -136,14 +138,11 @@ fn real_main() -> i32 {
 		}
 	}
 
+	let (logs_tx, logs_rx) = mpsc::channel::<LogEntry>();
```
Comment (antiochp) [inline src/bin/grin.rs]: I have not finished reading this PR fully but posting this here (mainly to remind me to look at this again) - this is an async channel. If the receiver is too slow for any reason then the queue will grow indefinitely.

File: util/src/types.rs [hunk 2]
```diff
@@ -1,104 +1 @@
...
-			tui_running: None,
-		}
-	}
-}
-
-use std::ops::Deref;
-use zeroize::Zeroize;
-/// Zeroing string, mainly useful for password
-#[derive(Clone, PartialEq, PartialOrd)]
-pub struct ZeroingString(String);
```
Comment (antiochp) [inline util/src/types.rs]: This seems unrelated? And its password related (masking out passwords in the logs?) So probably best to have this in a separate PR if this is indeed unused and to be cleaned up.

File: src/bin/tui/ui.rs [hunk 3]
```diff
@@ -124,6 +133,10 @@ impl UI {
 			return false;
 		}
 
+		while let Some(message) = self.logs_rx.try_iter().next() {
```
Comment (antiochp) [inline src/bin/tui/ui.rs]: We may want to rethink some of the tui refresh logic. I think we update the _entire_ tui on every step, regardless of what view or screen in the tui we are actually in.

Particularly for something as heavy as the full screen of log messages, we may want to look into limiting updates here only if we are actually viewing this screen. (Maybe this is a bad idea, I have not thought this through fully).

But it feels like updating many lines of streaming log msg output might be overly expensive if the tui is not actually viewing these at the time. Particularly if the node is logging debug (or trace...) level logs.

Comment (JosephGoulden) [reply src/bin/tui/ui.rs]: We do up date the whole thing every second yes. But we could update the logs view only when its the active tab as you suggest. When moving to a separate tab we would clear the logs view and consume but ignore the log messages. So when you went back to the logs view it would start scrolling up from the bottom with the latest messages.

File: util/src/types.rs (hunk 2 above)
Comment (JosephGoulden) [reply util/src/types.rs]: Yes it's unrelated and unused. Will put it back for now.

File: src/bin/grin.rs (hunk 1 above)
Comment (JosephGoulden) [reply src/bin/grin.rs]: Ok, could try the sync_channel instead with a reasonable buffer? I'm not sure how this would affect the sender (logging) threads if the buffer filled and they were blocked.

**Comment by JosephGoulden** [comment]
Thanks @antiochp and thanks for the review. I'll have a look at those issues you raised and wait for any further comments.

File: src/bin/grin.rs (hunk 1 above)
Comment (antiochp) [reply src/bin/grin.rs]: There's a `try_send` on a sync sender that we can take advantage of here.
Basically just drop log messages on the floor if we cannot safely send it via the buffer.

https://doc.rust-lang.org/std/sync/mpsc/struct.SyncSender.html#method.try_send

File: src/bin/tui/ui.rs (hunk 3 above)
Comment (antiochp) [reply src/bin/tui/ui.rs]: Might be worth getting some performance numbers and a rough idea of how costly this is to update before making these changes. Its possible this is not actually an issue and fine for now.

Comment (JosephGoulden) [reply src/bin/tui/ui.rs]: So I did some testing and observational measurement between this branch and master and I cannot discern any noticeable impact on resources or responsiveness of the TUI.

Typically grin uses between 10-20% CPU and ~23MB RAM on my machine and this didn't change with TUI trace logging enabled. Also tested with log to file and TUI logging enabled. And tested with chain sync from scratch, again there was no noticeable difference.

File: src/bin/grin.rs (hunk 1 above)
Comment (JosephGoulden) [reply src/bin/grin.rs]: done this now

File: src/bin/tui/ui.rs [hunk 4]
```diff
@@ -173,8 +187,8 @@ impl Controller {
 			while let Some(message) = self.rx.try_iter().next() {
 				match message {
 					ControllerMessage::Shutdown => {
+						warn!("Shutdown in progress, please wait");
 						self.ui.stop();
-						println!("Shutdown in progress, please wait");
```
[the remark this answers is not in the document]
Comment (JosephGoulden) [reply src/bin/tui/ui.rs]: yeah I don't like the text appearing randomly in the middle of the TUI, let me see if I can find a better way

[the remark this answers is not in the document]
Comment (hashmap) [reply src/bin/tui/ui.rs]: I hacked it, unfortunately could not find a way to show a dialog window there

File: servers/src/grin/server.rs [hunk 5]
```diff
@@ -80,38 +80,31 @@ pub struct Server {
...
-	pub fn start<F>(config: ServerConfig, mut info_callback: F) -> Result<(), Error>
-	where
-		F: FnMut(Server),
-	{
-		let mining_config = config.stratum_mining_config.clone();
-		let enable_test_miner = config.run_test_miner;
-		let test_miner_wallet_url = config.test_miner_wallet_url.clone();
-		let serv = Server::new(config)?;
+	/// Start ancillary services
+	pub fn start_ancillary(&self) -> Result<(), Error> {
```
[the remark this answers is not in the document]
Comment (JosephGoulden) [reply servers/src/grin/server.rs]: I changed the name because I moved the Server::new(config) out to the calling function, this is where the server is actually started.
And the reason I did that is because the Receiver is not clonable so has to be passed as another parameter in the callback which did not look very pretty.
I can do it this way if you want though, I don't really mind. Will change it so you can see.

File: src/bin/grin.rs [hunk 6]
```diff
@@ -136,14 +138,11 @@ fn real_main() -> i32 {
 		}
 	}
 
+	let (logs_tx, logs_rx) = mpsc::sync_channel::<LogEntry>(200);
```
[the remark this answers is not in the document]
Comment (JosephGoulden) [reply src/bin/grin.rs]: yeah I think I can do that, and wrap Sender and Receiver with Option.

File: src/bin/tui/ui.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (JosephGoulden) [reply src/bin/tui/ui.rs]: ok.. actually I'll revert this for now and see about doing it in a separate MR.

File: util/src/lib.rs [hunk 7]
```diff
@@ -42,8 +42,9 @@ pub use crate::logger::{init_logger, init_test_logger};
 pub mod secp_static;
 pub use crate::secp_static::static_secp_instance;
 
+/// Zeroing string for passwords
```
[the remark this answers is not in the document]
Comment (JosephGoulden) [reply util/src/lib.rs]: Fixed now. Accidentally removed the comments in util/src/types.rs

**Comment by JosephGoulden** [comment]
Thanks for the review @quentinlesceller . I've fixed all your suggestions, let me know what you think.

**Comment by quentinlesceller** [comment]
Currently looking at the changes now. Is there a way to wrap the text when reducing the windows size ?

**Comment by JosephGoulden** [comment]
> Currently looking at the changes now. Is there a way to wrap the text when reducing the windows size ?

Yes there is.. and I've now done this.

**Comment by quentinlesceller** [comment]
@JosephGoulden  Just tried and somehow the colors are not working anymore with the logs.
See attached screenshot (might be something with my terminal):
Capture d’écran 2019-10-30 à 14 07 38

**Comment by JosephGoulden** [comment]
@quentinlesceller  fixed that now.

File: util/src/types.rs [hunk 8]
```diff
@@ -12,64 +12,7 @@
...
-			file_log_level: LogLevel::Info,
-			log_file_path: String::from("grin.log"),
-			log_file_append: true,
-			log_max_size: Some(1024 * 1024 * 16), // 16 megabytes default
-			log_max_files: Some(DEFAULT_ROTATE_LOG_FILES),
-			tui_running: None,
-		}
-	}
-}
+//! Zeroing String
```
Comment (quentinlesceller) [inline util/src/types.rs]: Weird comment here?

Comment (JosephGoulden) [reply util/src/types.rs]: It's module documentation and it seems to be required for all public modules. Without it we get a warning due to the #![warn(missing_docs)] in lib.rs.

After I removed the logging code from this file I just updated the comment to say ZeroingString as that is all that is left in there. Maybe the file should be renamed to zeroing_string.rs. Or a better comment?

Comment (quentinlesceller) [reply util/src/types.rs]: Okay let's leave it like that for now.

**Comment by antiochp** [comment]
I think something may have slipped through here.

When we parse the config file we accept either `Info` or `INFO` as the log level.
Both appear to be parsed correctly.

But - when we _generate_ the config file we default to `INFO` which is not supported in previous code.
